### DISCUSSION TITLE
How can I re-render server components when data updates?

### DISCUSSION THREAD

**Discussion by FlyingTurkman** [vercel/next.js#56750] [Help]
### Summary

I want to re-render server components with mongodb change stream. Here is an example

```tsx
export default async function Page() {
    const cookieStore = cookies()
    const token = cookieStore.get('token')?.value
    let queues: queueType[] = await getQueues()
    if (queues.length > 0) {
        const queuesChangeStream = Queues.watch([], { fullDocument: 'updateLookup' })
        queuesChangeStream.on('change', async (changeEvent: ChangeStreamDocument) => {
            if (changeEvent.operationType == 'update' && changeEvent.fullDocument) {
                const newQueue: queueType = {
                    _id: changeEvent.fullDocument._id,
                    maxElo: changeEvent.fullDocument.maxElo,
                    minElo: changeEvent.fullDocument.minElo,
                    queueName: changeEvent.fullDocument.queueName,
                    queueUrl: changeEvent.fullDocument.queueUrl,
                    players: changeEvent.fullDocument.players
                }
                const oldQueues: queueType[] = queues.filter((q) => q._id.toString() != newQueue._id.toString())
                queues = oldQueues.concat([newQueue])
            }
        })
    }
    return(
        <div>
            <PageView queues={queues} token={token}/>
        </div>
    )
}
```

I want to re-render PageView component when queues updated.

**Comment by icyJoseph** [comment]
I think there was talk, in a podcast somewhere, about how a WebSockets, and likewise approaches, are not yet implemented/possible.

Here the flow would be a Server Pushing down to a client, or sort of emitting an event to the client. WebSockets, Server Sent Events, etc, would support this, but the actual flow we have is a Request arriving to a server and the server closing the socket after it is done, even if it streams the response. 

There's also the added issue of serverless deploys, and how the lambda handling the request might just be shutdown after a short time. Not applicable if you host this yourself in a non-serverless way.

Maybe if you host this yourself, you could use a Route handler that emits events to the client (Server Sent Events or WebSockets), and that client updates the UI based of that data?

- https://developer.mozilla.org/en-US/docs/Web/API/Server-sent_events/Using_server-sent_events  

Podcast, https://changelog.com/jsparty/267:

> But I know one of the things we’ve talked about – so for example, you can wait for a single piece of data to resolve. You can do an async/await, and – one of the things that we’re exploring is support for like an async iterable, so that you can actually await for a list of values to resolve, and then stream those down. So that kind of addresses it, but that is the sort of list of items use case, where you want to just like load one after the other, and kind of incrementally stream them in. If you’re talking about things like WebSockets, where you want to just have kind of like a Pub/Sub system that’s like sending you information, that right now would be kind of done as a client component.

**Reply by FlyingTurkman** [reply]
Looks like final solution web socket but I don't want to use it. I am looking for other solutions but I havent found it yet.
